### ISSUE TITLE
Different semantics on `atomicAllExch`

### ISSUE THREAD

**Issue by Flamefire** [ComputationalRadiationPhysics/picongpu#2601]
The optimized version might have a different semantics due to this line https://github.com/ComputationalRadiationPhysics/picongpu/blob/a4658f429a94350f4aa4a2f019a9d87dc621f038/include/pmacc/nvidia/atomic.hpp#L195

If I understand this correctly then the unoptimized version makes the old value visible to all threads, the optimized version just to the leader. This is at least unexpected and might lead to bugs.

**Comment by psychocoderHPC** [comment]
If we need the old value we need to broadcast the result of `atomicExch` with shuffle.

**Comment by Flamefire** [comment]
You are right, the name is misleading. So it rather is an `atomicAllSet`? Is it actually required then as in "doesn't CUDA already guarantee atomic writes on word level"? I think this is quite some overhead for the equivalent of `*ptr = value` where `T` is at most `double`.

**Comment by Flamefire** [comment]
> I do not understand your point.

You do a `*ptr = value` from multiple threads at the same time. In the case where there are multiple values involved, is it guaranteed that `*ptr` has exactly one of these? Counter-example where this is not the case: `T` is a struct with 2 doubles. There will be (at least) 2 writes (one for each value). With different threads trying to write different values this might lead to ultimately having the first and second value from different threads.

As a `double` consists of 8 bytes the same might happen when writing only 1 double value: Thread A writes the first 4 bytes and Thread B the second 4. This is what I mean with "word size". If the word size is 8 bytes and those are written at once, then this cannot happen.

> No if you wrote with different threads/warps to one address than you will have race conditions (this is what cuda-memcheck is saying)

Then this is a (kinda) false-positive: All threads write the same value. So even though it is technically a race condition there will be no damage. So question is: If all threads write the same value anyway, why can't the higher level code chose exactly one to write this value?
